### ISSUE TITLE
How to avoid Fatal Exception: java.lang.IllegalStateException Attempt to use disposed object

### ISSUE THREAD

**Issue by iTruff** [twilio/twilio-chat-demo-android#31]
How to avoid this exception? I can't reproduce it in debug mode, but I am receiving a lot of reports with such exceptions in Crashlytics. Maybe I am doing smth wrong?

```
Fatal Exception: java.lang.IllegalStateException
Attempt to use disposed object in Messages#getLastConsumedMessageIndex
com.twilio.chat.Messages.getLastConsumedMessageIndex
```

**Comment by berkus** [comment]
Hi, looks like the messages object is disposed but still some callbacks are getting through.

I'm currently revamping the lifecycle code in the SDK, because there are some races and not all native objects are disposed at the same time as the java ones - which could cause this behavior. Please stay tuned, I will release an update in some time.

Will keep this issue open until fixed.

**Comment by iTruff** [comment]
Thanks for answer.

**Comment by berkus** [comment]
Hi, I've started a new version of object lifecycle management in the SDK; hopefully will release it soon.

**Comment by iTruff** [comment]
Thanks you very much for feedback :)

**Comment by mcnarek** [comment]
Any news about this issue? @berkus

**Comment by mcnarek** [comment]
I am always getting "Attempt to use disposed object" error, will it be possible to check if the channels (or objects) is disposed, then use them? Thanks

**Comment by berkus** [comment]
I'm working on the improvement. Properly managing memory on the edge between Java and C++ is rather hard, so it takes time.

If you get "Attempt to use disposed object" errors then it means the object was destroyed and you couldn't use it anyway. Can you show some example code where it happens? Probably it could be done differently to avoid this situation.

**Comment by mcnarek** [comment]
` public void setData(final ChannelModel channelModel) {

            channelModel.getChannel(new CallbackListener<Channel>() {
                @Override
                public void onSuccess(final Channel channel) {
                    if (user != null) {
                        if (user.getUsername().equals(channel.getCreatedBy())) {
                            name.setText(channel.getFriendlyName());
                        } else {
                            name.setText(channel.getCreatedBy());
                        }
                    }
                    if (channelModel.getDateCreatedAsDate() != null) {
                        date.setText(DateFormatter.formatChatDate(channelModel.getDateCreatedAsDate().toString(), new SimpleDateFormat(PATTERN, Locale.ENGLISH)));
                    }
                    final Messages messagesObject = channel.getMessages();
                    if (messagesObject != null) {
                        messagesObject.getLastMessages(1, new CallbackListener<List<Message>>() {
                            @Override
                            public void onSuccess(final List<Message> messages) {
                                if (messages.size() > 0) {
                                    channel.getUnconsumedMessagesCount(new CallbackListener<Long>() { // here Attempt to use disposed object in Channel#getUnconsumedMessagesCount
                                        @Override
                                        public void onSuccess(final Long aLong) {
                                           ... 
                                        }
                                    });
                                    lastMessage.setText(messages.get(0).getMessageBody());
                                    if (messages.get(0).getTimeStampAsDate() != null) {
                                        date.setText(DateFormatter.formatChatDate(messages.get(0).getTimeStampAsDate().toString(),
                                            new SimpleDateFormat(PATTERN, Locale.ENGLISH)));
                                    } else {
                                        date.setText(DateFormatter.formatChatDate(channel.getDateCreatedAsDate().toString(),
                                                new SimpleDateFormat(PATTERN, Locale.ENGLISH)));
                                    }
                                } else {
                                    lastMessage.setText("");
                                    try {
                                        date.setText(DateFormatter.formatChatDate(channel.getDateCreatedAsDate().toString(),
                                                new SimpleDateFormat(PATTERN, Locale.ENGLISH)));
                                    } catch (final IllegalStateException ignored) {
                                    }
                                }
                            }
                        });
                    }
                }
            });
        }`

**Comment by mcnarek** [comment]
`    public void getChannel(final CallbackListener<Channel> listener) {

        if (channel != null) {
            listener.onSuccess(channel);
            return;
        }
        if (channelDescriptor != null) {
            channelDescriptor.getChannel(listener);
            return;
        }
        listener.onError(new ErrorInfo(-10005, "No channel in model"));
    }
`

**Comment by berkus** [comment]
Hmm, looks like this Channel is getting destroyed elsewhere.

Do you invoke `channel.dispose()` anywhere explicitly?

**Comment by mcnarek** [comment]
No I don't call `"channel.dispose()"` method from anywhere

but I do call `channel.destroy(listener)` method when I delete the chat channel
and also call `twilioChatClient.shutDown()` method when I log out the user.
Are these OK?

Thanks, for response.

**Comment by berkus** [comment]
The `twilioChatClient.shutdown()` will implicitly dispose all the Channels and Messages that were created from this chat client.

It seems that you're still using them after the shutdown, hence the given message. When you shutdown and then create a new client do you properly drop all the old references and fetch them from the new client again?

**Comment by berkus** [comment]
There's a version coming up, 3.0.1 which will relax checks on disposed objects and only log a warning instead of throwing exceptions.

**Comment by berkus** [comment]
3.0.1 has been released with this fix; however for 2.0 version of the SDK there's still new version pending release, so I'm not closing this one yet.
